src/Raven.Server/RavenServer.cs

```diff
@@ -929,45 +938,179 @@ internal AuthenticateConnection AuthenticateConnectionCertificate(X509Certificat
                 using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
                 using (ctx.OpenReadTransaction())
                 {
-                    var certKey = Constants.Certificates.Prefix + certificate.Thumbprint;
-                    var cert = ServerStore.Cluster.Read(ctx, certKey) ??
-                               ServerStore.Cluster.GetLocalState(ctx, certKey);
-
-                    if (cert == null)
-                    {
-                        authenticationStatus.Status = AuthenticationStatus.UnfamiliarCertificate;
-                    }
-                    else
+                    try
                     {
-                        var definition = JsonDeserializationServer.CertificateDefinition(cert);
-                        authenticationStatus.Definition = definition;
-                        if (definition.SecurityClearance == SecurityClearance.ClusterAdmin)
-                        {
-                            authenticationStatus.Status = AuthenticationStatus.ClusterAdmin;
-                        }
-                        else if (definition.SecurityClearance == SecurityClearance.ClusterNode)
-                        {
-                            authenticationStatus.Status = AuthenticationStatus.ClusterAdmin;
-                        }
-                        else if (definition.SecurityClearance == SecurityClearance.Operator)
+                        var cert = ServerStore.Cluster.GetCertificateByThumbprint(ctx, certificate.Thumbprint) ??
+                                   ServerStore.Cluster.GetLocalStateByThumbprint(ctx, certificate.Thumbprint);
+
+                        if (cert == null)
                         {
-                            authenticationStatus.Status = AuthenticationStatus.Operator;
+                            // If connection will be approved, cert won't be null anymore and will be assigned the relevant permissions
+                            MaybeAllowConnectionBasedOnPinningHash(certificate, ctx, ref authenticationStatus, ref cert, remoteAddress);
```

Comment:
This can be a pretty expensive operation. We read certificates, parse them, compute PK, etc.
I wonder if we should put some rate limiting here to prevent ourselves from being hammered ?

Reply:
Note, if we do, that should be in another PR

Reply:
I agree